dali/pipeline/graph_descr.cc

```diff
@@ -71,358 +73,436 @@ void CheckOpConstraints(const OpSpec &spec) {
+  for (auto &node : op_nodes_) {
     if (node.instance_name == name) {
       return node;
     }
   }
-  // Search gpu nodes
-  for (auto& node : gpu_nodes_) {
-    if (node.instance_name == name) {
-      return node;
+  DALI_FAIL("Operator node with name " + name + " not found.");
+}
+
+namespace {
+  std::ofstream& PrintTo(std::ofstream &ofs, const OpNode& node, bool show_ids) {
+    ofs << node.instance_name;
+    if (show_ids) {
+      ofs << "_" << node.id;
     }
+    return ofs;
   }
-  // Search mixed nodes
-  for (auto& node : mixed_nodes_) {
-    if (node.instance_name == name) {
-      return node;
+  std::ofstream&  PrintTo(std::ofstream &ofs, const TensorNode& node, bool show_ids) {
+    ofs << node.name;
+    if (show_ids) {
+      ofs << "_" << node.id;
     }
+    return ofs;
   }
-  // Search support nodes
-  for (auto& node : support_nodes_) {
-    if (node.instance_name == name) {
-      return node;
+  std::string GetOpColor(OpType op_type) {
+    switch (op_type) {
+      case OpType::CPU:
+        return "blue";
+      case OpType::GPU:
+        return "#76b900";
+      case OpType::MIXED:
+        return "cyan";
+      case OpType::SUPPORT:
+        return "grey";
+      default:
+        return "black";
     }
   }
-  DALI_FAIL("Operator node with name " + name + " not found.");
+}  // namespace
+
+void OpGraph::GenerateDOTFromGraph(std::ofstream &ofs, bool show_tensors, bool show_ids,
+                                   bool use_colors) {
+  // Just output all the edges
+  for (auto &op : op_nodes_) {
+    if (use_colors) {
+      PrintTo(ofs, op, show_ids) << "[color=\"" << GetOpColor(op.op_type) << "\"];\n";
+    }
+    for (auto child_id : op.children) {
+      auto& child_node = Node(child_id);
+      PrintTo(ofs, op, show_ids) << " -> ";
+      PrintTo(ofs, child_node, show_ids);
+      if (show_tensors) {
+        ofs << "[style=dotted]";
+      }
+      ofs << ";\n";
+    }
+    if (show_tensors) {
+      int i = 0;
+      for (auto t_id : op.children_tensors) {
+        TensorNode& child_tensor = Tensor(t_id);
+        PrintTo(ofs, child_tensor, show_ids) << "[shape=box];\n";
+        PrintTo(ofs, op, show_ids) << " -> ";
+        PrintTo(ofs, child_tensor, show_ids) << "[label=" << i++ <<"];\n";
+        GenerateDOTFromGraph(child_tensor, ofs, show_tensors, show_ids);
+      }
+    }
+    ofs << "\n";
+  }
+}
+
+void OpGraph::GenerateDOTFromGraph(const TensorNode &current_node, std::ofstream &ofs,
+                                   bool show_tensors, bool show_ids) {
+  for (auto edge : current_node.consumers) {
+      PrintTo(ofs, current_node, show_ids) << " -> ";
+      auto &child_op = Node(edge.node);
+      PrintTo(ofs, child_op, show_ids) << "[label=" << edge.index << "];\n";
+  }
 }
 
 template <>
 bool OpGraph::TensorIsType<CPUBackend>(const string &name) {
-  return TensorSourceMeta(name).is_cpu;
+  return TensorSourceMeta(name).storage_device == StorageDevice::CPU;
```

Comment:
You could make `is_cpu` a function that performs this test (not necessarily in this PR).